### ISSUE TITLE
"movement" effect broken on turn 1 ?

### ISSUE THREAD

**Issue by vgaming** [wesnoth/wesnoth#2374]
Steps to reproduce:

1. start wesnoth v1.13.10+dev (bbdb5ad8-Clean)
2. start a new Human vs Human game, without any add-ons or custom maps
3. On side 1 turn 1, execute this:
`wesnoth.add_modification(wesnoth.get_units{side=2}[1], "object", {{"effect", { apply_to="movement", increase=100 }}})`
If you disable fog, you'll see enemy leader has a movement like 105/105.
4. Now end turn and see moves for side 2.

Expected at this point: Unit of side 2 has movement 105/105 (hundred and something).

Actually: unit of side 2 will have movement like 5/105. The maximum changed, but the available value is low (without changes).

Note: this only happens at turn 1. If you do this at turn 2, everything will be fine.

**Comment by vgaming** [comment]
@ProditorMagnus indeed, wiki has the current behavior documented. Why is it desired though? I wonder if we should change the contract then. May be not though.

Thoughts?

**Comment by soliton-** [comment]
Healing and income make little sense on turn 1. Movement refresh might though. Is it difficult to do the refresh manually? If not a change is probably not worth considering.

**Comment by vn971** [comment]
> Is it difficult to do the refresh manually?

Nope, not really that hard. A Lua call to traverse all units at turn start and set `moves` to `max_moves` is about enough. I was just surprised why would unit have lower `moves` than it should have (definitely not a case of [least surprise](https://en.wikipedia.org/wiki/Principle_of_least_astonishment))).
But anyway, if it's deemed better to keep the situation as-is (for example not to break some backwards compat on rare cases), feel free to close the issue.

**Comment by vgaming** [comment]
> Is it difficult to do the refresh manually?

Nope, not really that hard. A Lua call to traverse all units at turn-side start and set `moves` to `max_moves` is about enough. I was just surprised why would unit have lower `moves` than it should have (definitely not a case of [least surprise](https://en.wikipedia.org/wiki/Principle_of_least_astonishment))).
But anyway, if it's deemed better to keep the situation as-is (for example not to break some backwards compat on rare cases), feel free to close the issue.

**Comment by soliton-** [comment]
Perhaps @ln-zookeeper has a comment on this, otherwise I‘ll close later.

**Comment by CelticMinstrel** [comment]
There was a proposal to add a WML tag (and underlying Lua call) to explicitly perform turn refresh actions, which would solve this issue (among others). I can't recall if these actions are neatly bundled into a function on the C++ side though; if not that would make things a little complicated...

**Comment by soliton-** [comment]
IIRC turn refresh is basically one function call.

The question here IMO is rather what to do by default though. vgaming mentioned what he wants can already be done fairly easily.

So to me the remaining question is whether it makes sense to refresh movement by default on turn 1 or not.

**Comment by GregoryLundberg** [comment]
I'm just jumping in here but my opinion: mark this invalid and close it.

Test 1: AToTB "prestart" event, add a lua block with the OP example code at the end of the event. Start the campaign and check Side 2 leader has the expected 105/105 once we clear the objectives. End turn and check again when you can, still 105/105 for side 2 leader.

Test 2: AToTB "start" event, move the Lua block from prestart to the very top of the start event. Again, start and clear the objectives and not the Side 2 leader has 105/105 End turn and check again when you can, still 105/105 for side 2 leader.

Test 3: AToTB "side 1 turn" event, create the event and move the Lua block into it.  Again, start and clear the objectives and not the Side 2 leader has 105/105 End turn and check again when you can, still 105/105 for side 2 leader.

Test 4: AToTB "side 1 turn end" event. As above, but also create a "side 2 turn 1" event and have it [inspect]. Finally, we get the reported error.

To my mind, this indicates no issues and the OP sees it because it's in debug mode and run by-hand.

If you put the modification in the WML at reasonable points, all is well.

It is possible to reproduce the issue, but you have to do it way later .. that is, in a "turn end" event.

**Comment by vgaming** [comment]
> the OP sees it because it's in debug mode and run by-hand.
> If you put the modification in the WML at reasonable points, all is well.

In my opinion, if that is indeed the case, it's a bug in bold text highlighted in red. Running a command via Lua should be no different than including something in a WML block. I'll try to confirm tests1-2-3 too.

**Comment by soliton-** [comment]
@vgaming Could you please follow up with your findings? I agree that it may be a bug if it works differently in the debug console.

**Comment by vgaming** [comment]
@soliton- sorry, yes, this somehow slipped through.

I tried to repro tests 1-4. I don't know what AToTB is, hope it's add-on name and not important. I've used a clean map and exactly one add-on with the following _main.cfg:
```
[modification]
	id=deleteme
	name="deleteme"
	description="deleteme"
	require_modification=no
	[event]
		name="prestart"
		[lua]
			code= <<
				wesnoth.add_modification(wesnoth.get_units{side=2}[1], "object", {{"effect", { apply_to="movement", increase=100 }}})
			>>
		[/lua]
	[/event]
[/modification]
```
I observe the same behavior I reported. When you end turn for side 1, you have side2 with movement that looks like 5/105 (the increase is applied to maximum, but not current value).
I tried the same test with "start" and "side 1 turn" event names, the result is the same.

Wesnoth version is v1.13.11 (bdd5089dd7-Clean).

@GregoryLundberg can you please explain what does AToTB mean, and whether your set-up was different from mine above (I leterally pasted the whole content of _main.cfg).

**Comment by CelticMinstrel** [comment]
@vgaming AToTB is "A Tale of Two Brothers", one of the mainline campaigns.

**Comment by gfgtdf** [comment]
per this comment https://github.com/wesnoth/wesnoth/blob/fe5296aefd91675692259e660ab0254e49763dce/src/play_controller.cpp#L498 this is now fixed and it now behaves as op expects it.

**Comment by gfgtdf** [comment]
edit wait nevermind this is about healing
